drivers/firmware/nrf_ironside/ironside_dvfs.c

```diff
@@ -0,0 +1,203 @@
+		return -EINVAL;
+	}
+
+	nrf_hsfll_trim_t hsfll_trim = {};
+	uint8_t freq_trim_idx = dvfs_hsfll_data[oppoint].new_f_trim_entry;
+
+#if defined(NRF_APPLICATION)
+	hsfll_trim.vsup = NRF_FICR->TRIM.APPLICATION.HSFLL.TRIM.VSUP;
+	hsfll_trim.coarse = NRF_FICR->TRIM.APPLICATION.HSFLL.TRIM.COARSE[freq_trim_idx];
+	hsfll_trim.fine = NRF_FICR->TRIM.APPLICATION.HSFLL.TRIM.FINE[freq_trim_idx];
+#if NRF_HSFLL_HAS_TCOEF_TRIM
+	hsfll_trim.tcoef = NRF_FICR->TRIM.APPLICATION.HSFLL.TRIM.TCOEF;
+#endif
+#else
+	/* Currently only application core is supported */
+	return -EINVAL;
+#endif
+
+	nrf_hsfll_clkctrl_mult_set(NRF_HSFLL, dvfs_hsfll_data[oppoint].new_f_mult);
+	nrf_hsfll_trim_set(NRF_HSFLL, &hsfll_trim);
+	nrf_barrier_w();
+
+	nrf_hsfll_task_trigger(NRF_HSFLL, NRF_HSFLL_TASK_FREQ_CHANGE);
+	/* Trigger hsfll task one more time, SEE PAC-4078 */
+	nrf_hsfll_task_trigger(NRF_HSFLL, NRF_HSFLL_TASK_FREQ_CHANGE);
+
+	return 0;
+}
+
+/* Function handling steps for DVFS oppoint change. */
+static int ironside_dvfs_prepare_to_scale(enum ironside_dvfs_oppoint freq_setting)
+{
+	int err = 0;
+
+	if (ironside_dvfs_is_downscaling(freq_setting)) {
+		err = ironside_dvfs_configure_hsfll(freq_setting);
+	}
+
+	return err;
+}
+
+/* Update MDK variable which is used by nrfx_coredep_delay_us (k_busy_wait). */
+static void ironside_dvfs_update_core_clock(enum ironside_dvfs_oppoint oppoint_freq)
+{
+	extern uint32_t SystemCoreClock;
+
+	SystemCoreClock = dvfs_hsfll_data[oppoint_freq].max_hsfll_freq;
+}
+
+/* Perform scaling finnish procedure. */
+static int ironside_dvfs_change_oppoint_complete(enum ironside_dvfs_oppoint dvfs_oppoint)
+{
+	int err = 0;
+
+	if (!ironside_dvfs_is_downscaling(dvfs_oppoint)) {
+		err = ironside_dvfs_configure_hsfll(dvfs_oppoint);
+	}
+
+	current_dvfs_oppoint = dvfs_oppoint;
+	ironside_dvfs_update_core_clock(dvfs_oppoint);
+
+	return err;
+}
+
+/**
+ * @brief Request DVFS oppoint change from IRONside secure domain.
+ * This function will send a request over IPC to the IRONside secure domain
+ * This function is synchronous and will return when the request is completed.
+ *
+ * @param oppoint @ref enum ironside_dvfs_oppoint
+ * @return int
+ */
+static int ironside_dvfs_req_oppoint(enum ironside_dvfs_oppoint oppoint)
+{
+	int err;
+
+	struct ironside_call_buf *const buf = ironside_call_alloc();
+
+	buf->id = IRONSIDE_CALL_ID_DVFS_SERVICE_V0;
+	buf->args[IRONSIDE_DVFS_SERVICE_OPPOINT_IDX] = oppoint;
+
+	ironside_call_dispatch(buf);
+
+	if (buf->status == IRONSIDE_CALL_STATUS_RSP_SUCCESS) {
+		err = buf->args[IRONSIDE_DVFS_SERVICE_RETCODE_IDX];
+	} else {
+		err = buf->status;
+	}
+
+	ironside_call_release(buf);
+
+	return err;
+}
+
+int ironside_dvfs_change_oppoint(enum ironside_dvfs_oppoint dvfs_oppoint)
+{
+	int status = 0;
+
+	if (!ironside_dvfs_opp_allowed(dvfs_oppoint)) {
+		return -EINVAL;
+	}
+
+	if (dvfs_oppoint == current_dvfs_oppoint) {
+		return 0;
+	}
+
+	if (!k_mutex_lock(&ironside_dvfs_mutex, K_MSEC(1000))) {
```

Comment:
Should this timeout of 1000ms be configurable? The default here seems quite high, I'd imagine that it would be nice to be able to tweak it.